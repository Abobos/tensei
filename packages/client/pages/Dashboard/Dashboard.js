import cn from 'classnames'
import React, { Fragment } from 'react'
import { Link, Route, Switch } from 'react-router-dom'
import {
    Dropdown,
    DropdownList,
    DropdownListItem,
    Paragraph,
} from '@contentful/forma-36-react-components'

import ShowResource from '~/pages/ShowResource'
import ResourceIndex from '~/pages/ResourceIndex'
import CreateResource from '~/pages/CreateResource'

import { withResources } from '~/store/resources'
import { mustBeAuthenticated } from '~/store/auth'

import ArrowIcon from '~/components/ArrowIcon'

class DashboardPage extends React.Component {
    state = this.defaultState()

    defaultState() {
        const resources = this.props.resources.filter(
            (resource) => resource.displayInNavigation
        )

        const groups = [
            {
                name: 'Dashboard',
                open: true,
                slug: 'dashboard',
                items: [
                    {
                        slug: 'home',
                        label: 'Home',
                        to: Tensei.getPath('dashboard'),
                    },
                ],
            },
        ]

        resources.forEach((resource) => {
            const navItem = {
                slug: resource.slug,
                to: Tensei.getPath(`resources/${resource.slug}`),
                label: resource.label,
            }

            const groupExists = groups.findIndex(
                (group) => group.name === resource.group
            )

            if (groupExists === -1) {
                groups.push({
                    open: true,
                    slug: resource.groupSlug,
                    name: resource.group,
                    items: [
                        {
                            slug: resource.slug,
                            label: resource.label,
                            to: Tensei.getPath(`resources/${resource.slug}`),
                        },
                    ],
                })
            } else {
                groups[groupExists] = {
                    ...groups[groupExists],
                    items: [
                        ...groups[groupExists].items,
                        {
                            slug: resource.slug,
                            label: resource.label,
                            to: Tensei.getPath(`resources/${resource.slug}`),
                        },
                    ],
                }
            }
        })

        return {
            groups,
            accountMenuOpen: false,
        }
    }

    logout = () => {
        Tensei.request.post('logout').then(() => {
            window.location.href = Tensei.getPath('auth/login')
        })
    }

    toggleAccountMenu = () => {
        this.setState({
            accountMenuOpen: !this.state.accountMenuOpen,
        })
    }

    toggleGroup = (groupToToggle) => {
        this.setState({
            groups: this.state.groups.map((group) => {
                if (group.slug === groupToToggle.slug) {
                    return {
                        ...group,
                        open: !group.open,
                    }
                }

                return group
            }),
        })
    }

    renderGroups = () => {
        return (
            <>
                {this.state.groups.map((group) => (
                    <Fragment key={group.slug}>
                        <header
                            onClick={() => this.toggleGroup(group)}
                            className="flex items-center border-b border-gray-lightest-100 p-3 cursor-pointer hover:bg-gray-lightest-100 transition duration-75 w-full uppercase text-blue-darkest-200 font-bold text-xs"
                        >
                            <ArrowIcon
                                className={cn(
                                    'mr-3 transition duration-75 transform text-blue-darkest',
                                    {
                                        '-rotate-90': !group.open,
                                    }
                                )}
                            />
                            {group.name}
                        </header>
                        {group.open &&
                            group.items.map((item) => (
                                <Link key={item.slug} to={item.to}>
                                    <div
                                        className={cn(
                                            'w-full py-3 pl-10 pr-3 text-sm cursor-pointer hover:bg-gray-lightest-200 transition duration-75',
                                            {
                                                'bg-gray-lightest-100': this.props.location.pathname.match(
                                                    item.to
                                                ),
                                            }
                                        )}
                                    >
                                        {item.label}
                                    </div>
                                </Link>
                            ))}
                    </Fragment>
                ))}
            </>
        )
    }

    render() {
        const { accountMenuOpen } = this.state
        return (
            <div className="w-full">
                <div className="w-full h-topbar bg-blue-darkest flex items-center text-white justify-between">
                    <span></span>
                    <span></span>
                    <div>
                        <Dropdown
                            isOpen={accountMenuOpen}
                            onClose={this.toggleAccountMenu}
                            isAutoalignmentEnabled={true}
                            toggleElement={
                                <div
                                    data-testid="dashboard-header-dropdown"
                                    onClick={this.toggleAccountMenu}
                                    className={cn(
                                        'shadow-account-menu cursor-pointer px-5 h-topbar flex items-center justify-center',
                                        {
                                            'bg-blue-darkest-200 hover:bg-blue-darkest-300': !accountMenuOpen,
                                            'bg-blue-darkest-300': accountMenuOpen,
                                        }
                                    )}
                                >
                                    <img
                                        className="w-8 h-8 rounded-full"
                                        src="https://avatars.githubusercontent.com/u/19477966?v=3"
                                        alt=""
                                    />
                                    <ArrowIcon className="ml-3 text-white w-4 h-4" />
                                </div>
                            }
                        >
                            <DropdownList data-testid="dashboard-header-dropdown-list">
                                <DropdownListItem onClick={console.log}>
                                    Account settings
                                </DropdownListItem>
                            </DropdownList>
                            <DropdownList
                                border="top"
                                data-testid="dashboard-header-dropdown-list"
                            >
                                <DropdownListItem onClick={this.logout}>
                                    Logout
                                </DropdownListItem>
                            </DropdownList>
                        </Dropdown>
                    </div>
                </div>

                <div
                    className="w-full flex flex-wrap"
                    style={{ height: 'calc(100vh - 4.375rem)' }}
                >
                    <div className="border-r border-gray-lighter h-full bg-gray-lightest w-full md:w-18-percent">
                        {this.renderGroups()}
                    </div>
                    <div className="w-full md:w-82-percent flex-grow h-full overflow-scroll">
                        <div className="py-4 px-3 pt-8 pb-12 md:px-12">
                            <Switch>
                                <Route
                                    exact
                                    path={Tensei.getPath('resources/:resource')}
                                    component={ResourceIndex}
                                />
                                <Route
                                    path={Tensei.getPath(
                                        'resources/:resource/new'
                                    )}
                                    exact
                                    component={CreateResource}
                                />
                                <Route
                                    path={Tensei.getPath(
                                        'resources/:resource/:resourceId'
                                    )}
                                    exact
                                    component={ShowResource}
                                />
                                <Route
                                    path={Tensei.getPath(
                                        'resources/:resource/:resourceId/edit'
                                    )}
                                    component={CreateResource}
                                />
                            </Switch>
                            <div className="w-full flex items-center justify-center py-3 mt-24">
                                <Paragraph variant="small">
                                    Â© Tensei Admin {new Date().getFullYear()}
                                </Paragraph>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

export default withResources(mustBeAuthenticated(DashboardPage))
