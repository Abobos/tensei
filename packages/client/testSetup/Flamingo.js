import TextField from '~/fields/Text'
import SelectField from '~/fields/Select'
import HasManyField from '~/fields/HasMany'
import DateTimeField from '~/fields/DateTime'
import TextareaField from '~/fields/Textarea'
import BelongsToField from '~/fields/BelongsTo'

export default {
    getPath: jest.fn(() => 'string'),
    fieldComponents: {
        TextField,
        SelectField,
        HasManyField,
        DateTimeField,
        TextareaField,
        BelongsToField,
        LinkField: TextField,
        NumberField: TextField,
        IntegerField: TextField,
        DateField: DateTimeField,
    },
    library: {
        Notification: { success: jest.fn(), error: jest.fn() },
    },
}
