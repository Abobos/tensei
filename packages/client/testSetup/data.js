export const resources = [
    {
        collection: 'posts',
        defaultPerPage: 1,
        displayInNavigation: true,
        fields: [
            {
                asObjectId: true,
                asString: false,
                attributes: {},
                component: 'TextField',
                creationRules: [],
                databaseField: 'name',
                defaultValue: '',
                description: '',
                fieldName: 'Text',
                inputName: 'name',
                isNullable: true,
                isRelationshipField: false,
                isSearchable: false,
                isSortable: false,
                isUnique: false,
                name: 'Name',
                rules: [],
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: true,
                sqlDatabaseFieldType: 'string',
                updateRules: [],
            },
            {
                asObjectId: true,
                asString: false,
                attributes: {},
                component: 'TextField',
                creationRules: [],
                databaseField: 'description',
                defaultValue: '',
                description: '',
                fieldName: 'Text',
                inputName: 'description',
                isNullable: true,
                isRelationshipField: false,
                isSearchable: false,
                isSortable: false,
                isUnique: false,
                name: 'description',
                rules: [],
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: true,
                sqlDatabaseFieldType: 'string',
                updateRules: [],
            },
            {
                asObjectId: true,
                asString: false,
                attributes: {},
                component: 'TextField',
                creationRules: [],
                databaseField: 'content',
                defaultValue: '',
                description: '',
                fieldName: 'Text',
                inputName: 'content',
                isNullable: true,
                isRelationshipField: false,
                isSearchable: false,
                isSortable: false,
                isUnique: false,
                name: 'content',
                rules: [],
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: true,
                sqlDatabaseFieldType: 'string',
                updateRules: [],
            },
        ],
        group: 'All',
        label: 'Posts',
        slug: 'posts',
        messages: {
            'title.required': 'The title field is required.',
            'publishedAt.required': 'The published at field is required.',
        },
        name: 'Post',
        param: 'posts',
        perPageOptions: [(1, 3, 5)],
        primaryKey: '_id',
    },
]

export const user = {
    email: 'dodo@email.com',
    firstName: 'dozie',
    lastName: 'nwoga',
    password: '$2a$10$d.IeGxbRR4kc1ZxE7u0LSuHMrX9aMlUrbLgLoxqEcVI9I2CyntgV.',
    _id: '5f0d62b4e2fab0431e1d35cf',
}
