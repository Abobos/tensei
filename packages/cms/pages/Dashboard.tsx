import Gravatar from 'react-gravatar'
import React, { useState } from 'react'
import { Transition, Menu } from '@tensei/components'
import { Route, Switch, Link } from 'react-router-dom'

import Nav from '../components/Nav'

import FourOhFour from './404'
import Settings from './Settings'
import ResourceIndex from './ResourceIndex'
import ResourceDetail from './ResourceDetail'
import CreateResource from './CreateResource'

export interface DashboardProps {}

const Dashboard: React.FC<DashboardProps> = () => {
    const [offCanvasOpen, setOffCanvasOpen] = useState(false)

    return (
        <>
            <div className="h-screen flex overflow-hidden bg-tensei-gray-100">
                {/* Off-canvas menu for mobile, show/hide based on off-canvas menu state. */}
                <Transition show={offCanvasOpen} className="md:hidden">
                    <div className="fixed inset-0 flex z-40">
                        <Transition.Child
                            enter="transition-opacity ease-linear duration-300"
                            enterFrom="opacity-0"
                            enterTo="opacity-100"
                            leave="transition-opacity ease-linear duration-300"
                            leaveFrom="opacity-100"
                            leaveTo="opacity-0"
                        >
                            {ref => (
                                <div
                                    ref={ref}
                                    className="fixed inset-0"
                                    aria-hidden={!offCanvasOpen}
                                >
                                    <div className="absolute inset-0 bg-gray-600 opacity-75" />
                                </div>
                            )}
                        </Transition.Child>
                        <Transition.Child
                            enter="transition ease-in-out duration-300 transform"
                            enterFrom="-translate-x-full"
                            enterTo="translate-x-0"
                            leave="transition ease-in-out duration-300 transform"
                            leaveFrom="translate-x-0"
                            leaveTo="-translate-x-full"
                        >
                            {ref => (
                                <div
                                    ref={ref}
                                    className="relative flex-1 flex flex-col max-w-xs w-full pt-5 pb-4 bg-tensei-darkest"
                                >
                                    <div className="absolute top-0 right-0 -mr-12 pt-2">
                                        <button
                                            onClick={() =>
                                                setOffCanvasOpen(!offCanvasOpen)
                                            }
                                            className="ml-1 flex items-center justify-center h-10 w-10 rounded-full focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white"
                                        >
                                            <span className="sr-only">
                                                Close sidebar
                                            </span>
                                            {/* Heroicon name: x */}
                                            <svg
                                                className="h-6 w-6 text-white"
                                                xmlns="http://www.w3.org/2000/svg"
                                                fill="none"
                                                viewBox="0 0 24 24"
                                                stroke="currentColor"
                                                aria-hidden="true"
                                            >
                                                <path
                                                    strokeLinecap="round"
                                                    strokeLinejoin="round"
                                                    strokeWidth={2}
                                                    d="M6 18L18 6M6 6l12 12"
                                                />
                                            </svg>
                                        </button>
                                    </div>
                                    <div className="flex-shrink-0 flex items-center px-4">
                                        <div className="flex items-center justify-start pl-6 h-12 rounded-lg bg-white w-full">
                                            <img
                                                className="h-6 w-auto"
                                                src="https://res.cloudinary.com/bahdcoder/image/upload/v1604236130/Asset_1_4x_fhcfyg.png"
                                                alt="Tensei"
                                            />
                                        </div>
                                    </div>
                                    <div className="mt-5 flex-1 h-0 overflow-y-auto">
                                        <Nav className="md:px-2 space-y-1" />
                                    </div>
                                </div>
                            )}
                        </Transition.Child>
                        <div className="flex-shrink-0 w-14" aria-hidden="true">
                            {/* Dummy element to force sidebar to shrink to fit close icon */}
                        </div>
                    </div>
                </Transition>
                {/* Static sidebar for desktop */}
                <div className="hidden bg-tensei-darkest md:flex md:flex-shrink-0">
                    <div className="flex flex-col w-64">
                        {/* Sidebar component, swap this element with another sidebar if you like */}
                        <div className="flex flex-col flex-grow pt-3 pb-4 overflow-y-auto">
                            <div className="flex items-center flex-shrink-0 px-4">
                                <div className="flex items-center justify-start pl-6 h-12 rounded-lg bg-white w-full">
                                    <img
                                        className="h-6 w-auto"
                                        src="https://res.cloudinary.com/bahdcoder/image/upload/v1604236130/Asset_1_4x_fhcfyg.png"
                                        alt="Tensei"
                                    />
                                </div>
                            </div>
                            <div className="mt-5 flex-1 flex flex-col">
                                <Nav className="flex-1 space-y-1 mt-3" />
                            </div>
                        </div>
                    </div>
                </div>
                <div className="flex flex-col w-0 flex-1 overflow-hidden">
                    <div className="relative z-10 flex-shrink-0 flex h-16 bg-white border-b border-tensei-gray-600">
                        <button
                            onClick={() => {
                                setOffCanvasOpen(!offCanvasOpen)
                            }}
                            className="px-4 border-r border-gray-200 text-gray-500 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-indigo-500 md:hidden"
                        >
                            <span className="sr-only">Open sidebar</span>
                            {/* Heroicon name: menu-alt-2 */}
                            <svg
                                className="h-6 w-6"
                                xmlns="http://www.w3.org/2000/svg"
                                fill="none"
                                viewBox="0 0 24 24"
                                stroke="currentColor"
                                aria-hidden="true"
                            >
                                <path
                                    strokeLinecap="round"
                                    strokeLinejoin="round"
                                    strokeWidth={2}
                                    d="M4 6h16M4 12h16M4 18h7"
                                />
                            </svg>
                        </button>
                        <div className="flex-1 px-8 flex justify-between">
                            <div className="flex-1 flex items-center">
                                <h2 className="text-tensei-darkest font-bold">
                                    {window.Tensei.state.config.name}
                                </h2>
                            </div>
                            <div className="ml-4 flex items-center md:ml-6">
                                {/* Profile dropdown */}
                                <div className="relative inline-block text-left">
                                    <Menu>
                                        {({ open }) => (
                                            <>
                                                <span className="rounded-md shadow-sm">
                                                    <Menu.Button
                                                        className={`max-w-xs bg-white flex items-center text-sm rounded-full focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-tensei-primary`}
                                                    >
                                                        <span className="sr-only">
                                                            Open user menu
                                                        </span>
                                                        <Gravatar
                                                            className="h-10 w-10 rounded-full"
                                                            email={
                                                                window.Tensei
                                                                    .state.admin
                                                                    .email
                                                            }
                                                        />
                                                    </Menu.Button>
                                                </span>

                                                <Transition
                                                    show={open}
                                                    enter="transition ease-out duration-100"
                                                    enterFrom="transform opacity-0 scale-95"
                                                    enterTo="transform opacity-100 scale-100"
                                                    leave="transition ease-in duration-75"
                                                    leaveFrom="transform opacity-100 scale-100"
                                                    leaveTo="transform opacity-0 scale-95"
                                                >
                                                    <Menu.Items
                                                        static
                                                        className="shadow absolute right-0 w-56 mt-2 origin-top-right bg-white border border-tensei-gray-600 rounded-lg shadow-sm outline-none"
                                                    >
                                                        <div className="px-4 py-3">
                                                            <p className="text-sm leading-5 text-tensei-gray-800">
                                                                Signed in as
                                                            </p>
                                                            <p className="font-medium leading-5 truncate">
                                                                {
                                                                    window
                                                                        .Tensei
                                                                        .state
                                                                        .admin
                                                                        .email
                                                                }
                                                            </p>
                                                        </div>

                                                        <div>
                                                            <Menu.Item>
                                                                {({
                                                                    active
                                                                }) => (
                                                                    <Link
                                                                        to={window.Tensei.getPath(
                                                                            'settings'
                                                                        )}
                                                                        className={`${
                                                                            active
                                                                                ? 'bg-tensei-gray-100'
                                                                                : ''
                                                                        } flex justify-between w-full px-4 py-3 leading-5 text-left`}
                                                                    >
                                                                        Settings
                                                                    </Link>
                                                                )}
                                                            </Menu.Item>
                                                        </div>

                                                        <div>
                                                            <Menu.Item>
                                                                {({
                                                                    active
                                                                }) => (
                                                                    <button
                                                                        onClick={() => {
                                                                            window.Tensei.client
                                                                                .post(
                                                                                    'logout'
                                                                                )
                                                                                .then(
                                                                                    () => {
                                                                                        window.location.href = window.Tensei.getPath(
                                                                                            'auth/login'
                                                                                        )
                                                                                    }
                                                                                )
                                                                        }}
                                                                        className={`${
                                                                            active
                                                                                ? 'bg-tensei-gray-100'
                                                                                : ''
                                                                        } flex justify-between w-full px-4 py-3 leading-5 text-left`}
                                                                    >
                                                                        Logout
                                                                    </button>
                                                                )}
                                                            </Menu.Item>
                                                        </div>
                                                    </Menu.Items>
                                                </Transition>
                                            </>
                                        )}
                                    </Menu>
                                </div>
                            </div>
                        </div>
                    </div>
                    <main
                        tabIndex={0}
                        className="flex-1 relative overflow-y-auto overflow-x-hidden focus:outline-none"
                    >
                        <Switch>
                            <Route
                                exact
                                component={ResourceIndex}
                                path={window.Tensei.getPath(
                                    'resources/:resource'
                                )}
                            />
                            <Route
                                exact
                                component={CreateResource}
                                path={window.Tensei.getPath(
                                    'resources/:resource/create'
                                )}
                            />
                            <Route
                                exact
                                component={CreateResource}
                                path={window.Tensei.getPath(
                                    'resources/:resource/:id/update'
                                )}
                            />
                            <Route
                                exact
                                component={ResourceDetail}
                                path={window.Tensei.getPath(
                                    'resources/:resource/:id'
                                )}
                            />

                            <Route
                                component={Settings}
                                path={window.Tensei.getPath('settings')}
                            />

                            <Route
                                component={FourOhFour}
                                path={window.Tensei.getPath('404')}
                            />
                            {window.Tensei.ctx.routes.map(route => (
                                <Route
                                    key={route.path}
                                    path={route.path}
                                    exact={route.exact}
                                    component={route.component}
                                />
                            ))}
                            <Route component={FourOhFour} />
                        </Switch>
                    </main>
                </div>
            </div>
        </>
    )
}

export default Dashboard
