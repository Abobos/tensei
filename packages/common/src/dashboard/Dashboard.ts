import { paramCase } from 'change-case'
import {
  CardContract,
  DashboardConfig,
  SerializedDashboardContract,
  DashboardContract
} from '@tensei/common'

export class Dashboard implements DashboardContract {
  public config: DashboardConfig = {
    name: '',
    slug: '',
    cards: [],
    group: 'Dashboards',
    displayInNavigation: true
  }

  constructor(name: string, slug?: string) {
    this.config.name = name
    this.config.slug = slug || paramCase(name)
  }

  cards(cards: CardContract[]) {
    this.config.cards = cards

    return this
  }

  group(group: string) {
    this.config.group = group

    return this
  }

  public serialize(): SerializedDashboardContract {
    return {
      name: this.config.name,
      slug: this.config.slug,
      group: this.config.group,
      groupSlug: paramCase(this.config.group),
      displayInNavigation: this.config.displayInNavigation,
      cards: this.config.cards.map(card => card.serialize())
    }
  }

  public hideFromNavigation() {
    this.config.displayInNavigation = false

    return this
  }
}

export const dashboard = (name: string, slug?: string) =>
  new Dashboard(name, slug)
