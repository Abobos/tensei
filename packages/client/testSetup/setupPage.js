import React from 'react'
import { render } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { MemoryRouter } from 'react-router-dom'

import Auth from '~/store/auth'
import Resources from '~/store/resources'
import { resources, user } from './data'

const history = createMemoryHistory()

const setupPage = (
    initialEntries,
    initialIndex,
    Component,
    withAuth = true
) => {
    const props = { resources }
    return render(
        <MemoryRouter
            initialIndex={initialIndex}
            initialEntries={initialEntries}
        >
            <Auth.Provider value={withAuth ? [user, () => jest.fn()] : []}>
                <Resources.Provider
                    value={{
                        resources: resources,
                    }}
                >
                    <Component {...props} location={history.location} />
                </Resources.Provider>
            </Auth.Provider>
        </MemoryRouter>
    )
}
export default setupPage
