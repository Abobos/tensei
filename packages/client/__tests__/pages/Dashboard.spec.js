import React from 'react'
import { cleanup, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import { createMemoryHistory } from 'history'
import { MemoryRouter } from 'react-router-dom'

import Auth from '~/store/auth'
import Resources from '~/store/resources'
import Dashboard from '~/pages/Dashboard'
import { resources, user } from '~/testSetup/data'

const history = createMemoryHistory()

console.error = jest.fn()

const DashBoardSetup = () => {
    const props = { resources }
    return (
        <MemoryRouter
            initialIndex={0}
            initialEntries={[
                '/resources/posts',
                '/resources/posts/new',
                'auth/login',
            ]}
        >
            <Auth.Provider
                value={{
                    user,
                    authorizedToCreate: jest.fn(() => true),
                    authorizedToUpdate: jest.fn(() => true),
                    authorizedToDelete: jest.fn(() => true),
                }}
            >
                <Resources.Provider
                    value={{
                        resources: resources,
                    }}
                >
                    <Dashboard {...props} location={history.location} />
                </Resources.Provider>
            </Auth.Provider>
        </MemoryRouter>
    )
}

describe('Test the dashboard page', () => {
    beforeEach(() => {
        window.Tensei = {
            getPath: jest.fn(() => 'string'),
        }
    })

    afterEach(cleanup)

    test('should match snapshot', () => {
        const { asFragment } = render(<DashBoardSetup />)

        expect(asFragment()).toMatchSnapshot()
    })

    test('resources should be listed on the sidebar', () => {
        render(<DashBoardSetup />)

        resources.map((r) => {
            expect(screen.getByText(r.label)).toBeInTheDocument()
        })
    })

    test('resources can be toggled to hide / show', () => {
        render(<DashBoardSetup />)

        resources.map((r) => {
            expect(screen.getByText(r.label)).toBeInTheDocument()
        })

        userEvent.click(screen.getByText('Resources'))

        resources.map((r) => {
            expect(screen.queryByText(r.label)).not.toBeInTheDocument()
        })
    })

    test('the dashboard header has the appropriate links', () => {
        render(<DashBoardSetup />)

        userEvent.click(screen.getByTestId('dashboard-header-dropdown'))

        expect(
            screen.getAllByTestId('dashboard-header-dropdown-list')
        ).toHaveLength(2)
        expect(screen.getByText(/Account settings/i)).toBeInTheDocument()
        expect(screen.getByText(/Logout/i)).toBeInTheDocument()

        userEvent.click(screen.getByTestId('dashboard-header-dropdown'))

        expect(screen.queryByText(/Account settings/i)).not.toBeInTheDocument()
        expect(screen.queryByText(/Logout/i)).not.toBeInTheDocument()
    })

    test('clicking logout on the header dropdown should log user out', () => {
        window.Tensei = {
            getPath: jest.fn(() => 'auth/login'),
            request: { post: jest.fn(() => Promise.resolve(true)) },
            location: { assign: jest.fn() },
        }

        render(<DashBoardSetup />)

        userEvent.click(screen.getByTestId('dashboard-header-dropdown'))
        userEvent.click(screen.getByText(/Logout/i))

        expect(window.Tensei.request.post).toHaveBeenCalledWith('logout')
    })
})
