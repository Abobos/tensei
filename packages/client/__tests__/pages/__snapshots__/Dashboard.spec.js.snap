// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the dashboard page should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <div
      class="w-full h-topbar bg-blue-darkest flex items-center text-white justify-between"
    >
      <span />
      <span />
      <div>
        <div
          class="Dropdown__Dropdown___nAsJ-"
          data-test-id="cf-ui-dropdown"
          position="bottom-left"
        >
          <div
            aria-expanded="false"
            aria-haspopup="menu"
            class="shadow-account-menu cursor-pointer px-5 h-topbar flex items-center justify-center bg-blue-darkest-200 hover:bg-blue-darkest-300"
            data-testid="dashboard-header-dropdown"
          >
            <img
              alt=""
              class="w-8 h-8 rounded-full"
              src="https://avatars.githubusercontent.com/u/19477966?v=3"
            />
            <svg
              class="w-4 h-4 ml-3 text-white w-4 h-4"
              data-testid="ArrowIcon"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 9l-7 7-7-7"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full flex flex-wrap"
      style="height: calc(100vh - 4.375rem);"
    >
      <div
        class="border-r border-gray-lighter h-full bg-gray-lightest w-full md:w-18-percent"
      >
        <header
          class="flex items-center border-b border-gray-lightest-100 p-3 cursor-pointer hover:bg-gray-lightest-100 transition duration-75 w-full uppercase text-blue-darkest-200 font-bold text-xs"
        >
          <svg
            class="w-4 h-4 mr-3 transition duration-75 transform text-blue-darkest"
            data-testid="ArrowIcon"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
          Dashboard
        </header>
        <a
          href="/resources/string"
        >
          <div
            class="w-full py-3 pl-10 pr-3 text-sm cursor-pointer hover:bg-gray-lightest-200 transition duration-75"
          >
            Home
          </div>
        </a>
        <header
          class="flex items-center border-b border-gray-lightest-100 p-3 cursor-pointer hover:bg-gray-lightest-100 transition duration-75 w-full uppercase text-blue-darkest-200 font-bold text-xs"
        >
          <svg
            class="w-4 h-4 mr-3 transition duration-75 transform text-blue-darkest"
            data-testid="ArrowIcon"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
          Resources
        </header>
        <a
          href="/resources/string"
        >
          <div
            class="w-full py-3 pl-10 pr-3 text-sm cursor-pointer hover:bg-gray-lightest-200 transition duration-75"
          >
            Posts
          </div>
        </a>
        <a
          href="/resources/string"
        >
          <div
            class="w-full py-3 pl-10 pr-3 text-sm cursor-pointer hover:bg-gray-lightest-200 transition duration-75"
          >
            News
          </div>
        </a>
      </div>
      <div
        class="w-full md:w-82-percent flex-grow h-full overflow-scroll"
      >
        <div
          class="py-4 px-3 pt-8 pb-12 md:px-12"
        >
          <div
            class="w-full flex items-center justify-center py-3 mt-24"
          >
            <p
              class="Paragraph__Paragraph___2aO-9"
              data-test-id="cf-ui-paragraph"
              variant="small"
            >
              Â© Flamingo Admin 2020
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
